### ISSUE TITLE
[java] New rule: UnnecessaryBoxing

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#2973]
**Proposed Rule Name:** UnnecessaryBoxing (pmd7)

The rule "PrimitiveWrapperInstantiation" has been implemented for PMD 6.37.0.

**Proposed Category:** Best Practices

**Description:** 
The new rules that merges together:
* [UnnecessaryWrapperObjectCreation](https://pmd.github.io/latest/pmd_rules_java_performance.html#unnecessarywrapperobjectcreation) -> UnnecessaryBoxing: #3364 
* BooleanInstantiation --> PrimitiveWrapperInstantiation: #3365 
* ByteInstantiation -> PrimitiveWrapperInstantiation
* IntegerInstantiation -> PrimitiveWrapperInstantiation
* LongInstantiation -> PrimitiveWrapperInstantiation
* ShortInstantiation -> PrimitiveWrapperInstantiation

(all in category performance)

The first rule flags `valueOf` calls, and only cares about primitive wrappers that are created and unboxed explicitly in the same expression, like `Integer.valueOf(0).intValue()`. This is contrived and no one does it: the rule has zero violations in checkstyle, spring, or the jdk.

The other rules flag uses of the constructors of primitive wrappers (but for some reason not `Double`, `Character` or `Float`). The rationale is that using `valueOf` may avoid creating a new instance. This is true, but also,
* the constructors are deprecated since Java 9, because of project Valhalla
* usage of both the constructors and `valueOf` can be replaced nearly everywhere by autoboxing

So the idea is to merge all those into a new `UnnecessaryBoxing` rule that cares about:
- [x]    Explicit boxing in contexts where the primitive would be autoboxed: `Object i = Integer.valueOf(2)`
- [x]    Explicit unboxing in a context where it would be auto-unboxed: `int i = integer.intValue()`
- [x]    Explicit unboxing in contexts where the value is immediately reboxed: `Object i = integer.intValue()` (would rebox the int)
- [x]    Explicit boxing in contexts where the value is immediately unboxed: `int i = Integer.valueOf(0)` (would unbox the integer)
- [x]    Boxing of an already boxed value: `Integer.valueOf(integer)` (unboxes integer, then boxes it)
- [x]    Uses of `Integer.valueOf(someString)` where an int is expected: `Integer.parseInt` is more appropriate. -> #3500
- [x]    ~Casts that box/unbox their operand (`if ((boolean) Boolean.TRUE) ...`) are unnecessary~ -> this is left to rule UnnecessaryCast, see #3218 

The new rule `PrimitiveWrapperConstructor` cares about:
- [x]   Usage of primitive wrapper constructors, e.g. `new Integer(1)`. Instead, `Integer.valueOf(1)` should be used.

This applies to all primitives.

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
class Scratch {

    public static void main(String[] args) {

        Integer anInteger = 2; // ok
        
        Object a = Integer.valueOf(2);  // explicit boxing where the value would be autoboxed
        int b = anInteger.intValue();     // explicit unboxing where the value would be auto-unboxed
        Object c = anInteger.intValue();  // unboxing where the value is immediately reboxed
        int i = Integer.valueOf(0);     // boxing where the value is immediately unboxed
        
        Integer.valueOf(anInteger);       // boxing of already boxed value
        
        if ((boolean) Boolean.TRUE);    // Casts that unboxes the operand        
        
        // Actually a cast to a wrapper can be treated just like a valueOf call
        // (when the cast operand is primitive) and be handled exactly like a valueOf call,
        // so be handled in all the situations above.
        
        Integer o = (Integer) a; // a is an Object so the cast is not redundant
    }
}

```

**Possible Properties:**

None that I can think of.

**Comment by alexotoo** [comment]
Please correct below, integer should be uppercase Integer.intValue() and Integer.intValue() respectively. 
 int b = integer.intValue();     // explicit unboxing where the value would be auto-unboxed
 Object c = integer.intValue();  // unboxing where the value is immediately reboxed

**Comment by oowekyala** [comment]
`intValue` is an instance method called on a variable called `integer` (of type `Integer`, capital `I`), declared at the start, so the lowercase is correct

**Comment by alexotoo** [comment]
> `intValue` is an instance method called on a variable called `integer` (of type `Integer`, capital `I`), declared at the start, so the lowercase is correct. I'll rename the var for clarity

Gotcha,  clearer now.

**Comment by jsotuyod** [comment]
@oowekyala shouldn't we close this ticket? there seems to be nothing left to do…

**Comment by adangel** [comment]
It is finished for PMD 7, yes. But we tend to not close such rule issues, as people don't understand that it's only finished for PMD 7 and not for PMD 6 (they assume any new release of PMD contains all closed issues). So it has to stay open until we finally release PMD 7.

**Comment by oowekyala** [comment]
> as people don't understand that it's only finished for PMD 7 and not for PMD 6 (they assume any new release of PMD contains all closed issues).

We use milestones to categorize issues so I think we should assume that users can use milestones too/ read release notes :) Usually the reason we've kept some issues from pmd 7 open is when they could also be implemented on master. We use a needs-backport tag in that case, which is missing here. In the case of these rules, I don't think we should encourage anyone to implement them on master, so IMO it could be closed.

**Comment by adangel** [comment]
The new rule "UnnecessaryBoxing" has been implemented for PMD 7 via #3364 .

The rule "PrimitiveWrapperInstantiation" has been implemented already for PMD 6.37.0 via #3365.
